backend/src/test/java/org/cryptomator/hub/RemoteUserPullerTest.java

```diff
@@ -0,0 +1,277 @@
+package org.cryptomator.hub;
+
+import io.quarkus.test.junit.QuarkusTest;
+import org.cryptomator.hub.entities.Authority;
+import org.cryptomator.hub.entities.Group;
+import org.cryptomator.hub.entities.User;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.converter.ArgumentConversionException;
+import org.junit.jupiter.params.converter.ConvertWith;
+import org.junit.jupiter.params.converter.SimpleArgumentConverter;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.mockito.Mockito;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@QuarkusTest
+class RemoteUserPullerTest {
+
+	private RemoteUserPuller remoteUserPuller;
+	private RemoteUserProvider remoteUserProvider = Mockito.mock(RemoteUserProvider.class);
+	private User user = Mockito.mock(User.class);
+
+	@BeforeEach
+	void setUp() {
+		remoteUserPuller = new RemoteUserPuller();
+		remoteUserPuller.remoteUserProvider = remoteUserProvider;
+	}
+
+	@Nested
+	@DisplayName("Test sync users")
+	public class TestUser {
+
+		@DisplayName("test add users")
+		@ParameterizedTest(name = "KCUsers: {0} DBUsers: {1} AddedUsers: {2}")
+		@CsvSource(value = {
+				"foo,bar,baz;,;foo,bar,baz",
+				"foo,bar,baz;la,li,lu;foo,bar,baz",
+				"foo,bar,baz;la,bar,baz;foo",
+				"baz;foo,bar,baz;,",
+				",;foo,bar,baz;,",
+				",;,;,"
+		}, delimiterString = ";")
+		public void testAddUsers(@ConvertWith(StringArrayConverter.class) String[] keycloakUserIdString, @ConvertWith(StringArrayConverter.class) String[] databaseUserIdString, @ConvertWith(StringArrayConverter.class) String[] addedUserIdString) {
+			Map<String, User> keycloakUsers = Mockito.mock(Map.class);
+			Map<String, User> databaseUsers = Mockito.mock(Map.class);
```

Comment:
wouldn't it be easier to create Spy objects around "real" maps?

Reply:
Hmm I'm not sure, it's right that we don't want to test the map type itself but I find the initialization of the keys and values in the map easier and more comprehensible but I can change it if you want :)